File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    static int ret = 55;
    static int[] visited = new int[52];
    public void dfs(String start, String end, String[] words, int cnt){
        if (start.equals(end)){
            if(ret > cnt) ret = cnt;
            return;
        }
        if (cnt > ret) return;
        
        for (int i=0; i<words.length; i++){
            if (visited[i] == 0){
                int temp =0;
                for (int j=0; j<start.length(); j++){
                    if (start.charAt(j) != words[i].charAt(j))temp++;
                }
                if (temp == 1){
                    visited[i] = 1;
                    dfs(words[i],end, words, cnt+1);
                    visited[i] = 0;
                }
            }
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        dfs(begin,target,words,0);
        if (ret != 55)answer = ret;
        return answer;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int minCnt = 52;
bool visited[52];

void dfs(string begin, string target, vector<string> words, int cnt){
    if(cnt >= minCnt)return;
    if(begin == target){
        minCnt = minCnt < cnt ? minCnt: cnt;
        return;
    }
    
    for (int i=0; i<words.size(); i++){
        if(visited[i])continue;
        
        int charCnt = 0;
        for (int j = 0; j<begin.length(); j++){
            if(charCnt > 1)break;
            if(begin[j] != words[i][j]){
                charCnt++;
            }
        }
        if(charCnt == 1){
            visited[i] = true;
            dfs(words[i], target, words, cnt+1);
            visited[i] = false;
        }
    }
    
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    dfs(begin,target,words, 0);
    if (minCnt != 52) answer = minCnt;
    return answer;
}